consensus/core/src/context.rs

```diff
@@ -88,3 +96,35 @@ impl Context {
         self
     }
 }
+
+/// A clock that allows to derive the current UNIX system timestamp while guaranteeing that
+/// timestamp will be monotonically incremented having tolerance to ntp and system clock changes and corrections.
+/// Explicitly avoid to make `[Clock]` cloneable to ensure that a single instance is shared behind an `[Arc]`
+/// wherever is needed in order to make sure that consecutive calls to receive the system timestamp
+/// will remain monotonically increasing.
+pub(crate) struct Clock {
+    unix_epoch_instant: Instant,
+}
+
+impl Clock {
+    pub fn new() -> Self {
+        let now = Instant::now();
+        let duration_since_unix_epoch =
+            match SystemTime::now().duration_since(SystemTime::UNIX_EPOCH) {
+                Ok(d) => d,
+                Err(e) => panic!("SystemTime before UNIX EPOCH! {e}"),
+            };
+        let unix_epoch_instant = now.checked_sub(duration_since_unix_epoch).unwrap();
+
+        Self { unix_epoch_instant }
+    }
+
+    // Returns the current time expressed as UNIX timestamp in milliseconds.
+    // Calculated with Rust Instant to ensure monotonicity.
+    pub(crate) fn timestamp_utc_ms(&self) -> BlockTimestampMs {
```

Comment:
I've explicitly didn't move the `BlockTimestampMs` as I still believe the type should remain in the `block.rs` . If you believe otherwise please let me know